analysis/src/cratesio.rs

```diff
@@ -0,0 +1,93 @@
+//! This module abstracts the communication with crates.io for a given crate
+
+// TODO: A cheaper way to interact with crates.io can be working with their
+// experimental database dump that is updated daily, https://crates.io/data-access,
+// which will enable us to avoid making http requests and dealing with rate limits
+
+use crates_io_api::{SyncClient, Error};
+use anyhow::Result;
+use tabled::Tabled;
+
+#[derive(Tabled, Default)]
+pub struct CratesioReport<'a> {
+    pub name: &'a str,
+    pub is_hosted: bool,
+    pub downloads: u64,
+    pub dependents: u64,
+}
+
+pub struct CratesioAnalyzer;
+
+impl CratesioAnalyzer {
+    pub fn analyze_cratesio<'a>(&self, name: &'a str) -> Result<CratesioReport<'a>> {
+        let client = SyncClient::new(
+            //TODO: replace repo URL with an email address to receive traffic notification
+            "User-Agent: Whackadep (https://github.com/diem/whackadep)",
+            std::time::Duration::from_millis(1000),
+        )?;
+
+        let is_hosted = match self.is_hosted(&client, name) {
+            Ok(success) => success,
+            Err(error) => return Err(error.into()),
+        };
+
+        if !is_hosted {
+            let cratesio_report = CratesioReport {
+                name,
+                is_hosted,
+                ..Default::default()
+            };
+            return Ok(cratesio_report);
+        }
+
+        // TODO we are calling get_crate twice in this function, redundant
+        let crate_info = match client.get_crate(name) {
+            Ok(info) => info.crate_data,
+            Err(error) => return Err(error.into()),
+        };
+
+        let downloads = crate_info.downloads;
+
+        println!("fetching {}", name);
+
+        // TODO: crates_io_api makes unnecessary request to page through all dependents,
+        // therefore, taking a long time
+        // when the total count is actually present in each result page
+        // make a PR to the upstream crate? or write custom http request by ourselves?
+        let dependents = match self.get_total_dependents(&client, name){
+            Ok(total) => total,
+            Err(error) => return Err(error.into()),
+        };
```

Comment:
Anytime you find yourself with this code pattern you should covert it to use `?`